### PR TITLE
Issue #231: Decorators can read tokens from cookies

### PR DESCRIPTION
**Pull request by jvllmr**
My approach for looking after the token in places which aren't the header

### CODE REVIEW

**Comment by jvllmr** [comment]
Worked when testing with ReactJS

File: flask_praetorian/base.py [hunk 1]
```diff
@@ -721,21 +721,29 @@ def read_token_from_cookie(self):
...
+                token = getattr(self, "read_token_from_{place}".format(place=place.lower()))()
+                if token:
+                    break
+            except MissingToken as err:
+                error=err
+            except AttributeError:
+                flask.current_app.logger.warning("Flask_Praetorian hasn't implemented reading JWT tokens from location {place}. Please reconfigure JWT_PLACES.".format(place=place.lower()))
+                flask.current_app.logger.warning("Values accepted in JWT_PLACES are: {def_places}".format(def_places=DEFAULT_JWT_PLACES))
+
+        if not token and error:
```
Comment (dusktreader) [inline flask_praetorian/base.py]: I think it's sufficient here to just check for token to be set. If token is falsey, this should throw the MissingToken exception.

File: flask_praetorian/base.py [hunk 2]
```diff
@@ -721,21 +721,29 @@ def read_token_from_cookie(self):
...
+            raise MissingToken("Could not find {cookie_name} in any of the given locations: {locations}" \
```
Comment (dusktreader) [inline flask_praetorian/base.py]: escaping newlines is not preferred.  

```suggestion
            raise MissingToken("Could not find {cookie_name} in any of the given locations: {locations}".format(
                cookie_name=self.cookie_name,
                locations=self.jwt_places,
            ))  
```

Comment (dusktreader) [reply flask_praetorian/base.py]: Also, it doesn't make sense to include cookie_name if the client app isn't using cookies

**Review by dusktreader — CHANGES_REQUESTED** [review]

**Comment by jvllmr** [comment]
I have applied some changes and removed the if statement at the end. The token gets returned immediately in the for loop and if the for loop is escaped without returning before a MissingToken exception is raised. Also I replaced cookie_name with token. I know exception pass isn't good practice but I couldn't find a better solution right now.

**Comment by jvllmr** [comment]
The method now doesn't contain any if statement at all.

**Review by jvllmr — COMMENTED** [review]
This contains a draft for #257 although you think it would create constraints to the user object. I did this because this implementation doesn't add constraints as far as I could think about it.

**Review by jvllmr — COMMENTED** [review]
I think this implementation should be cleaner than my previous one. Only the except pass isn't good to look at.

**Review by jvllmr — COMMENTED** [review]
Oh and I forgot to change the test accordingly

File: flask_praetorian/base.py [hunk 3]
```diff
@@ -722,24 +723,27 @@ def read_token_from_cookie(self):
...
         """
-        Tries to unpack the token from the current flask request in the locations configured with JWT_PLACES
+        Tries to unpack the token from the current flask request
+        in the locations configured by JWT_PLACES.
         Check-Order is defined by the value order in JWT_PLACES.
         """
         for place in self.jwt_places:
             try:
-                return getattr(self, "read_token_from_{place}".format(place=place.lower()))()
+                return getattr(self, "read_token_from_{place}"
```
Comment (dusktreader) [inline flask_praetorian/base.py]: When wrapping argument lines, please put the first argument on a new-line

File: flask_praetorian/decorators.py [hunk 4]
```diff
@@ -86,7 +87,8 @@ def wrapper(*args, **kwargs):
                 not current_guard().roles_disabled,
                 "This feature is not available because roles are disabled",
             )
-            role_set = set([str(n.value) if isinstance(n, Enum) else str(n) for n in required_rolenames])
+            role_set = set([str(n.value) if isinstance(n, Enum)
```
Comment (dusktreader) [inline flask_praetorian/decorators.py:90]: I'm not sure why this line and others in this module were altered with a different style

File: flask_praetorian/base.py [hunk 5]
```diff
@@ -722,24 +723,27 @@ def read_token_from_cookie(self):
...
-                flask.current_app.logger.warning("Flask_Praetorian hasn't implemented reading JWT tokens from location {place}. Please reconfigure JWT_PLACES.".format(place=place.lower()))
-                flask.current_app.logger.warning("Values accepted in JWT_PLACES are: {def_places}".format(def_places=DEFAULT_JWT_PLACES))
-
-        raise MissingToken("Could not find token in any of the given locations: {locations}".format(
-            locations=self.jwt_places,
-        ))  
-        
-        
-        
+                current_app.logger.warning("Flask_Praetorian hasn't implemented reading JWT tokens from location {place}. Please reconfigure JWT_PLACES."
```
Comment (dusktreader) [inline flask_praetorian/base.py]: instead of importing current_app separately, please just use `flask.current_app`

**Review by dusktreader — CHANGES_REQUESTED** [review]

File: flask_praetorian/base.py (hunk 5 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: Will do

File: flask_praetorian/decorators.py (hunk 4 above)
Comment (jvllmr) [reply flask_praetorian/decorators.py:90]: The idea behind it was that the list comprehension is more readable when the else and the if statement are split into two lines. I can turn it into a one-liner again if you want.

File: flask_praetorian/base.py (hunk 3 above)
Comment (jvllmr) [reply flask_praetorian/base.py]: I will adjust it to your wrapping scheme. Line 749 and 740 are longer than 79 characters. Would you prefer this

```Python
#And what about making flake8 ignore the length of the line?
raise MissingToken(
            "Could not find token in any of the given locations: {locations}".format(  # noqa: E501
                locations=self.jwt_places)
                )
```
or this ?
``` Python
raise MissingToken(
            "Could not find token in any of the "
	    "given locations: {locations}".format( 
                locations=self.jwt_places)
                )
```

Comment (dusktreader) [reply flask_praetorian/base.py]: In the next minor version of flask-praetorian, I will be using black as a code formatter. My preferred style is:
```python
raise MissingToken(
    "Could not find token in any of the given locations: {locations}".format(
        locations=self.jwt_places,
    )
)
```

File: flask_praetorian/decorators.py (hunk 4 above)
Comment (dusktreader) [reply flask_praetorian/decorators.py:90]: When breaking it up into multiple lines to improve readability, I prefer:
```python
 role_set = set([
    str(n.value) if isinstance(n, Enum)
    else str(n)
    for n in required_rolenames
])
```
However, for these I think the readability on one line is fine

**Pull request closed without merging**